ext/drbbdup/drbbdup.c

```diff
@@ -1590,11 +1589,14 @@ drbbdup_instrument_dups(void *drcontext, void *tag, instrlist_t *bb, instr_t *in
             }
         }
         drreg_restore_all(drcontext, bb, instr);
-    } else if (drbbdup_is_at_end(instr) &&
-               (!is_last_special || drbbdup_ilist_has_unending_emulation(bb))) {
-        drreg_restore_all(drcontext, bb, instr);
-    } else if (drbbdup_is_at_end(instr) || drbbdup_is_exit_jmp_emulation_marker(instr)) {
-        /* Ignore instruction: hide drbbdup's own markers and the rest of the end. */
+    } else if (drbbdup_is_at_end(instr)) {
+        /* i#5906: if the emulation start label is missing we might still need to restore
+         * registers for blocks that don't end in a branch or for rep-expanded blocks. */
+        if (!is_last_special || drbbdup_ilist_has_unending_emulation(bb))
+            drreg_restore_all(drcontext, bb, instr);
```

Comment:
This still feels fragile: if we add some other case that doesn't trigger drbbdup_is_at_end_initial() it seems easy to forget to, or not realize we have to, update here, and then have more subtle problems requiring complex debugging.  Did you have a reason to not want to take me up on my suggestion of having a boolean `restored_all` that we set to true when we call `drreg_restore_all` for `drbbdup_is_at_end_initial`, and then here for `drbbdup_is_at_end` we call `drreg_restore_all` if `!restored_all`?  That makes the code clearer IMHO (makes it clear we **always** want to restore-all, which is not clear here with these conditions) and less fragile.

Reply:
The way the original code was written, I thought there might be cases where we intentionally don't call drreg_restore_all

I have updated it now.